app/controllers/application_controller.rb

```diff
@@ -5,6 +5,83 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
   end
 
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'users/create'
+    end
+  end
+
+  post '/signup' do
+    @user = User.new(username: params[:username], email: params[:email], password: params[:password])
+
+    if @user.save
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      @message = "Oops, something didn't work.  Try again."
+      redirect "/signup"
+    end
+  end
+
+  get '/tweets' do
+    if logged_in?
+      @tweets = Tweet.all
+
+      erb :'tweets/index'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do 
+    @user = User.find_by(username: params[:username])
+
+    if @user && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/logout' do
+    if logged_in?
+      logout
+    else
+      redirect '/login'
+    end
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:user_id]
+    end
+
+    def current_user
+      User.find(session[:user_id])
```

Comment:
Thanks for the link!

On Sun, Apr 10, 2016 at 10:41 AM Zac <notifications@github.com> wrote:

> In app/controllers/application_controller.rb
> <https://github.com/learn-co-students/sinatra-fwitter-group-project-v-000/pull/74#discussion_r59135184>
> :
>
> > +
> > +  get '/logout' do
> > +    if logged_in?
> > +      logout
> > +    else
> > +      redirect '/login'
> > +    end
> > +  end
> > +
> > +  helpers do
> > +    def logged_in?
> > +      !!session[:user_id]
> > +    end
> > +
> > +    def current_user
> > +      User.find(session[:user_id])
>
> That makes perfect sense. Oddly enough, I just watched a short tutorial
> about memoization the other day. https://gorails.com/episodes/memoization
> <http://url> Your explanation of why we use it works great for me
> conceptually. Thanks!
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/learn-co-students/sinatra-fwitter-group-project-v-000/pull/74/files/ef0b5ea9d62c791ff65a4d5a1b0bfac92de44654#r59135184>
>